socketio/src/asynchronous/client/client.rs

```diff
@@ -401,13 +484,31 @@ impl Client {
         Ok(())
     }
 
+    /// Indicates whether the client should try to reconnect
+    pub(crate) async fn should_reconnect(&self) -> bool {
+        let manually_disconnected = self.manually_disconnected.load(Ordering::Acquire);
```

Comment:
I think the disconnection handling can be modelled with a tokio channel and a dedicated enum, this would get rid of some of the complexity introduced by the `Arc<AtomicBool>`s which already lead to a small bug which @1c3t3a caught in this PR. 

So essentially you would send over the channel  in line 463 and 236 and matching on the receive here in 489. The enum would essentially have two members now, Manual and Server (e.g). I hope this makes sense to you, but i consider this the more safe and readable solution

Reply:
Thanks for the input! I'll think about the channels and get back to you in a bit. Can you tell me what bug was found with the use of `AtomicBool`? I don't recall it being mentioned earlier so I'm curious.

Reply:
just that you used Ordering::Relaxed instead of Ordering::Release on line 236

Reply:
About channels, I'm not clear on how they would help here. I do agree that replacing the AtomicBools with an enum would make it more readable. So how about something like this:

```rust
pub(crate) async fn should_reconnect(&self) -> bool {
    match self.disconnect_reason.read().await {
      DisconnectReason::Manual => false,
      DisconnectReason::Server => self.builder.read().await.reconnect_on_disconnect
    }
}
```

Reply:
> just that you used Ordering::Relaxed instead of Ordering::Release on line 236

Ahh, right. Yes, that was unintentional but I totally get your point. Accidents are easy with them.

Reply:
Perhaps the enum should also have something like `NetworkError` because that is an expected failure.

Reply:
Well, I presume you would have to wrap the enum still in an Arc for example, and there would have to be either a Option around your enum or an NoDisconnection (e.g.) variant to model that there was no disconnect. A channel would in that case just be empty. You also don't have to wrap it in any other concurrency primitives like an Arc (or other) that should be handled for you. Also no need to reset the channel, since you only handle the message inside

Reply:
Hmm, channel `Receiver`'s can't be cloned (at least for `mpsc` and `broadcast`). I'm not sure if there's another type of channel that makes sense for this particular case. Well, there's `watch` but that doesn't allow the `Sender` to be cloned. :smile:

It feels like trying to solve this with a channel will not allow us to escape an `Arc`. Do you know if there's another way to do this? My experience with channels isn't extensive, so it is entirely possible that I'm missing something obvious.

Reply:
Even if I wrap a receiver in an Arc, I would need to lock it because it would have to be mutable. I could do all that, but at the end of the day we will still have some synchronization primitive. I don't see any way to escape that because `Client` has to be cloneable. The conclusion I'm reaching is that a channel doesn't really add a lot of value and it seems like a really heavy way to get around having one additional enum value. That being said, I'll defer the final decision to you and @1c3t3a. Let me know how you'd like me to go about it. Thanks!

Reply:
You can look into `broadcast`: https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html, which allows multiple senders and multiple receivers and doesn't require mutable references for sending and receiving.

Reply:
But I agree that the underlying primitives don't get much better. Honestly this is a general problem of this crate since we only create one socket and then use it everywhere, which is not ideal.

Using the `AtomicBool` is equivalent with other code in this crate, however as @Colin1860 rightfully pointed out it is not ideal and easy to shoot ourselves in the foot here.

I would leave the final decision of what to do up to you, or @Colin1860 if he feels strongly about using channels :)

Reply:
Ok, I have replaced both of the `AtomicBool`'s with a `DisconnectReason` enum. I think the code looks much nicer now and is easier to follow. It still needs an `Arc<RwLock>` which is unavoidable.

Even [broadcast::Receiver::recv](https://docs.rs/tokio/1.28.2/tokio/sync/broadcast/struct.Receiver.html#method.recv) needs `&mut self` so I'm not sure it would have made things significantly better in this case.

Reply:
Fair. Lets stick to the atomic bools!